File: app-java/src/main/java/tech/thdev/java_udemy_sample/async/ImageDownloadThread.java
package tech.thdev.java_udemy_sample.async;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.DrawableRes;
import android.text.TextUtils;
import android.util.LruCache;
import android.widget.ImageView;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.net.URL;
import java.net.URLConnection;

/**
 * Created by Tae-hwan on 26/10/2016.
 */

public class ImageDownloadThread {

    /**
     * LruCache를 사용하여 Image를 캐쉬처리
     */
    private LruCache<String, WeakReference<Bitmap>> cache = new LruCache<>(5 * 1024 * 1024); // 5MiB


    private static ImageDownloadThread imageDownloadThread;

    private ImageDownloadThread() {

    }

    public static ImageDownloadThread getInstance() {
        if (imageDownloadThread == null) {
            synchronized (ImageDownloadThread.class) {
                if (imageDownloadThread == null) {
                    imageDownloadThread = new ImageDownloadThread();
                }
            }
        }
        return imageDownloadThread;
    }

    public void loadImage(@DrawableRes int loadImage, ImageView imageView, String url) {
        imageView.setImageResource(loadImage);
        if (cache.get(url) == null || cache.get(url).get() == null) {
            imageView.setTag(url);
            new Thread(new DownloadThread(imageView, url)).start();
        } else {
            imageView.setImageBitmap(cache.get(url).get());
        }
    }

    private class DownloadThread implements Runnable {

        private WeakReference<ImageView> weakReferenceImageView;

        private String resourceUrl;

        private URLConnection connection;
        private InputStream inputStream;
        private BufferedInputStream bufferedInputStream;

        public DownloadThread(ImageView imageView, String resourceUrl) {
            this.weakReferenceImageView = new WeakReference<>(imageView);
            this.resourceUrl = resourceUrl;
        }

        @Override
        public void run() {
            try {
                URL url = new URL(resourceUrl);
                connection = url.openConnection();
                connection.connect();

                inputStream = connection.getInputStream();
                bufferedInputStream = new BufferedInputStream(inputStream);

                cache.put(resourceUrl, new WeakReference<>(BitmapFactory.decodeStream(bufferedInputStream)));
                draw(resourceUrl);

            } catch (IOException e) {
                closeStream();

            } finally {
                closeStream();
            }
        }

        private void closeStream() {
            try {
                if (inputStream != null) {
                    inputStream.close();
                    inputStream = null;
                }
                if (bufferedInputStream != null) {
                    bufferedInputStream.close();
                    bufferedInputStream = null;
                }
            } catch (IOException e1) {
                e1.printStackTrace();
            }
        }

        private void draw(final String resourceUrl) {
            new Handler(Looper.getMainLooper()).post(new Runnable() {
                @Override
                public void run() {
                    ImageView imageView = weakReferenceImageView.get();
                    if (!TextUtils.isEmpty(resourceUrl) &&
                            imageView.getTag() != null &&
                            imageView.getTag().equals(resourceUrl) &&
                            cache.get(resourceUrl) != null &&
                            cache.get(resourceUrl).get() != null) {
                        imageView.setImageBitmap(cache.get(resourceUrl).get());
                    }
                }
            });
        }
    }
}

File: app-kotlin/src/main/java/tech/thdev/kotlin_udemy_sample/network/image/ImageDownloadThread.kt
package tech.thdev.kotlin_udemy_sample.network.image

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.util.LruCache
import android.widget.ImageView
import androidx.annotation.DrawableRes

import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.lang.ref.WeakReference
import java.net.URL
import java.net.URLConnection

/**
 * Created by Tae-hwan on 26/10/2016.
 */

object ImageDownloadThread {

    /**
     * LruCache를 사용하여 Image를 캐쉬처리
     */
    private val cache = LruCache<String, WeakReference<Bitmap>>(5 * 1024 * 1024) // 5MiB

    fun loadImage(@DrawableRes loadImage: Int, imageView: ImageView, url: String?) {
        imageView.setImageResource(loadImage)

        url?.let {
            if (cache.get(it) == null || cache.get(it).get() == null) {
                imageView.tag = it
                Thread(DownloadThread(imageView, it)).start()

            } else {
                imageView.setImageBitmap(cache.get(it).get())
            }
        }
    }

    private class DownloadThread(imageView: ImageView, private val resourceUrl: String) : Runnable {

        private val weakReferenceImageView: WeakReference<ImageView>

        private var connection: URLConnection? = null
        private var inputStream: InputStream? = null
        private var bufferedInputStream: BufferedInputStream? = null

        init {
            this.weakReferenceImageView = WeakReference(imageView)
        }

        override fun run() {
            try {
                val url = URL(resourceUrl)
                connection = url.openConnection()
                connection?.connect()

                inputStream = connection?.inputStream
                bufferedInputStream = BufferedInputStream(inputStream)

                cache.put(resourceUrl, WeakReference(BitmapFactory.decodeStream(bufferedInputStream)))
                draw(resourceUrl)

            } catch (e: IOException) {
                closeStream()

            } finally {
                closeStream()
            }
        }

        private fun closeStream() {
            try {
                if (inputStream != null) {
                    inputStream?.close()
                    inputStream = null
                }
                if (bufferedInputStream != null) {
                    bufferedInputStream?.close()
                    bufferedInputStream = null
                }
            } catch (e1: IOException) {
                e1.printStackTrace()
            }

        }

        private fun draw(resourceUrl: String) {
            Handler(Looper.getMainLooper()).post {
                val imageView = weakReferenceImageView.get()
                if (!TextUtils.isEmpty(resourceUrl) &&
                        imageView?.tag != null &&
                        imageView.tag == resourceUrl &&
                        cache.get(resourceUrl) != null &&
                        cache.get(resourceUrl).get() != null) {
                    imageView.setImageBitmap(cache.get(resourceUrl).get())
                }
            }
        }
    }
}
